pkg/server/observer.go

```diff
@@ -289,124 +289,166 @@ func (s *ObserverServer) GetFlows(
 	req *observer.GetFlowsRequest,
 	server observer.Observer_GetFlowsServer,
 ) (err error) {
-	return getFlowsFromObserver(req, server, s)
+	return getFlows(req, server, s)
 }
 
-func getFlowsFromObserver(
+func getFlows(
 	req *observer.GetFlowsRequest,
 	server observer.Observer_GetFlowsServer,
 	obs Observer,
 ) (err error) {
-	reply, err := getFlows(server.Context(), obs.GetLogger(), obs.GetRingBuffer(), req)
+	start := time.Now()
+	log := obs.GetLogger()
+	ring := obs.GetRingBuffer()
+
+	i := uint64(0)
+	defer func() {
+		size := ring.Cap()
+		took := time.Now().Sub(start)
+		log.Debug(
+			"GetFlows finished",
+			zap.Uint64("number_of_flows", i),
+			zap.Uint64("buffer_size", size),
+			zap.String("whitelist", logFilters(req.Whitelist)),
+			zap.String("blacklist", logFilters(req.Blacklist)),
+			zap.Duration("took", took),
+		)
+	}()
+
+	ringReader, err := newRingReader(server.Context(), ring, req)
 	if err != nil {
+		if err == io.EOF {
+			return nil
+		}
 		return err
 	}
-	for {
-		select {
-		case <-server.Context().Done():
-			return nil
-		case rep, ok := <-reply:
-			if !ok {
+	flowsReader, err := newFlowsReader(ringReader, req, log)
+	if err != nil {
+		return err
+	}
+
+	for ; ; i++ {
+		flow, err := flowsReader.Next(server.Context())
+		if err != nil {
+			if err == io.EOF {
 				return nil
 			}
-			err := server.Send(&observer.GetFlowsResponse{
-				ResponseTypes: &observer.GetFlowsResponse_Flow{
-					Flow: rep,
-				},
-			})
-			if err != nil {
-				return err
-			}
+			return err
+		}
+		err = server.Send(&observer.GetFlowsResponse{
+			ResponseTypes: &observer.GetFlowsResponse_Flow{
+				Flow: flow,
+			},
+		})
+		if err != nil {
+			return err
 		}
 	}
 }
 
-func getUntil(req *observer.GetFlowsRequest, defaultTime *types.Timestamp) (time.Time, error) {
-	until := req.GetUntil()
-	if until == nil {
-		until = defaultTime
-	}
-	return types.TimestampFromProto(until)
+type flowsReader struct {
+	ringReader           *container.RingReader
+	whitelist, blacklist filters.FilterFuncs
+	maxFlows             uint64
+	follow, timeRange    bool
+	flowsCount           uint64
+	start, end           time.Time
 }
 
-func getBufferCh(ctx context.Context, ring *container.Ring, req *observer.GetFlowsRequest) (ch <-chan *v1.Event, stop context.CancelFunc, err error) {
-	stop = func() {}
-
-	// s.ring.ReadFrom reads the values up to the last written index, i.e.,
-	// it will read all values from the given interval:
-	// [ lastWrite, s.ring.write [
-	lastWrite := ring.LastWriteParallel() + 1
-	readIdx := lastWrite - req.Number
+func newFlowsReader(r *container.RingReader, req *observer.GetFlowsRequest, log *zap.Logger) (*flowsReader, error) {
+	whitelist, err := filters.BuildFilterList(req.Whitelist)
+	if err != nil {
+		return nil, err
+	}
+	blacklist, err := filters.BuildFilterList(req.Blacklist)
+	if err != nil {
+		return nil, err
+	}
+	log.Debug("filters", zap.String("req", fmt.Sprintf("%+v", req)))
```

Comment:
nit: maybe combine into a single log

```
log.Debug("processing getflows request", 
    zap.Any("req", req),
    zap.Any("whitelist", whitelist),
    zap.Any("blacklist", blacklist"))
```

Reply:
`newFlowsReader` just creates a new `flowsReader` and doesn't actually process anything. Nonetheless, you have a point: I haven't updated the debug log to reflect that.